tls/extensions/s2n_server_alpn.c

```diff
@@ -66,7 +66,6 @@ static int s2n_alpn_recv(struct s2n_connection *conn, struct s2n_stuffer *extens
 
     uint8_t protocol_len = 0;
     POSIX_GUARD(s2n_stuffer_read_uint8(extension, &protocol_len));
-    POSIX_ENSURE_LT(protocol_len, s2n_array_len(conn->application_protocol));
 
```

Comment:
This now leads to [a compiler error](https://github.com/aws/s2n-tls/actions/runs/10955926981/job/30420850551?pr=4789):
```
   /home/runner/work/s2n-tls/s2n-tls/tls/extensions/s2n_server_alpn.c:69:5: error: result of comparison of constant 256 with expression of type 'uint8_t' (aka 'unsigned char') is always true [-Werror,-Wtautological-constant-out-of-range-compare]
     69 |     POSIX_ENSURE_LT(protocol_len, s2n_array_len(conn->application_protocol));
        |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /home/runner/work/s2n-tls/s2n-tls/utils/s2n_safety_macros.h:261:80: note: expanded from macro 'POSIX_ENSURE_LT'
    261 | #define POSIX_ENSURE_LT(a, b)                                 __S2N_ENSURE((a) < (b), POSIX_BAIL(S2N_ERR_SAFETY))
        |                                                               ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  /home/runner/work/s2n-tls/s2n-tls/utils/s2n_ensure.h:35:15: note: expanded from macro '__S2N_ENSURE'
     35 |         if (!(cond)) {             \
        |               ^~~~
  1 error generated.
```
I added a test instead.

Reply:
hmm... i wonder how we can preserve this check, though. in theory the `conn->application_protocol` size could be reduced and then we'd have an overwrite. this check would at least avoid that issue.

Reply:
The new unit test I added should actually enforce that. The test I wrote sends an alpn based on the size of conn->application_protocol, but then it checks that the final result matches the known max size, independent of the size of conn->application_protocol: https://github.com/aws/s2n-tls/blob/c0a43a1876c3ba99cd07e06259b6fe557e6b80f1/tests/unit/s2n_server_alpn_extension_test.c#L143-L145

So if conn->application_protocol shrinks, that test fails.